File: apps/cli-go/internal/session/session.go
package session

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"time"

	"github.com/nkootstra/xpose/internal/protocol"
)

const sessionFileName = "session.json"

// configDirOverride is set during tests to avoid polluting the real config.
var configDirOverride string

// TunnelEntry records one tunnel's identity so it can be resumed.
type TunnelEntry struct {
	Subdomain string `json:"subdomain"`
	Port      int    `json:"port"`
	Domain    string `json:"domain"`
}

// Session is the on-disk representation of a resumable tunnel session.
type Session struct {
	Tunnels   []TunnelEntry `json:"tunnels"`
	CreatedAt time.Time     `json:"createdAt"`
}

// configDir returns the xpose config directory, creating it if necessary when create is true.
func configDir(create bool) (string, error) {
	var dir string
	if configDirOverride != "" {
		dir = configDirOverride
	} else {
		base, err := os.UserConfigDir()
		if err != nil {
			return "", err
		}
		dir = filepath.Join(base, "xpose")
	}
	if create {
		if err := os.MkdirAll(dir, 0o755); err != nil {
			return "", err
		}
	}
	return dir, nil
}

// sessionPath returns the full path to the session file.
func sessionPath(create bool) (string, error) {
	dir, err := configDir(create)
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, sessionFileName), nil
}

// Save writes the session to disk.
func Save(s *Session) error {
	path, err := sessionPath(true)
	if err != nil {
		return err
	}
	data, err := json.MarshalIndent(s, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0o644)
}

// Load reads the session from disk. Returns nil (with no error) if the file
// does not exist or the session has expired.
func Load() (*Session, error) {
	path, err := sessionPath(false)
	if err != nil {
		return nil, err
	}

	data, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}

	var s Session
	if err := json.Unmarshal(data, &s); err != nil {
		return nil, nil // treat corrupt file as missing
	}

	window := time.Duration(protocol.SessionResumeWindowSeconds) * time.Second
	if time.Since(s.CreatedAt) > window {
		return nil, nil // expired
	}

	return &s, nil
}

// Clear removes the session file.
func Clear() error {
	path, err := sessionPath(false)
	if err != nil {
		return err
	}
	err = os.Remove(path)
	if errors.Is(err, os.ErrNotExist) {
		return nil
	}
	return err
}

File: packages/tunnel-core/src/session.ts
import { readFileSync, writeFileSync, mkdirSync, unlinkSync } from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";
import { PROTOCOL, type TunnelConfig } from "@xpose/protocol";

const SESSION_FILE_NAME = "session.json";

export interface TunnelEntry {
  subdomain: string;
  port: number;
  domain: string;
  /** Optional tunnel access-control and response configuration. */
  config?: TunnelConfig;
}

export interface Session {
  tunnels: TunnelEntry[];
  createdAt: string; // ISO 8601
}

/** Allow tests to override the config directory. */
let configDirOverride: string | undefined;

export function setConfigDirOverride(dir: string | undefined): void {
  configDirOverride = dir;
}

function getConfigDir(create: boolean): string {
  const dir = configDirOverride ?? join(homedir(), ".config", "xpose");
  if (create) {
    mkdirSync(dir, { recursive: true });
  }
  return dir;
}

function getSessionPath(create: boolean): string {
  return join(getConfigDir(create), SESSION_FILE_NAME);
}

/** Save a session to disk. */
export function saveSession(session: Session): void {
  const path = getSessionPath(true);
  writeFileSync(path, JSON.stringify(session, null, 2), "utf-8");
}

/**
 * Load a session from disk.
 * Returns `null` if the file does not exist, is corrupt, or has expired.
 */
export function loadSession(): Session | null {
  const path = getSessionPath(false);

  let data: string;
  try {
    data = readFileSync(path, "utf-8");
  } catch {
    return null; // file doesn't exist
  }

  let session: Session;
  try {
    session = JSON.parse(data);
  } catch {
    return null; // corrupt file
  }

  if (!session.createdAt || !Array.isArray(session.tunnels)) {
    return null;
  }

  const elapsed = Date.now() - new Date(session.createdAt).getTime();
  const windowMs = PROTOCOL.SESSION_RESUME_WINDOW_SECONDS * 1000;
  if (elapsed > windowMs) {
    return null; // expired
  }

  return session;
}

/** Remove the session file. */
export function clearSession(): void {
  try {
    unlinkSync(getSessionPath(false));
  } catch {
    // ignore — file may not exist
  }
}
